feature/networkinstance/policybasedvrfselection/ate_tests/protocol_dscp_rules_for_vrf_selection_test/protocol_dscp_rules_for_vrf_selection_test.go

```diff
@@ -0,0 +1,490 @@
+		p.AppendRule(rule)
+	}
+	return &pf
+}
+
+// getAllFlows returns a map of all Flows. Various tests use a subset of these flows.
+func getAllFlows(args *testArgs) map[string]*ondatra.Flow {
+
+	ateInterfaces := args.top.Interfaces()
+	dstEndPointVlan10 := ateInterfaces["atePort2Vlan10"]
+	dstEndPointVlan20 := ateInterfaces["atePort2Vlan20"]
+	dstEndPointVlan30 := ateInterfaces["atePort2Vlan30"]
+
+	var f = map[string]*ondatra.Flow{
+		"ipinipd10": getIPinIPFlow(args, dstEndPointVlan10, "ipinipd10", 10),
+		"ipinipd20": getIPinIPFlow(args, dstEndPointVlan20, "ipinipd20", 20),
+		"ipinipd30": getIPinIPFlow(args, dstEndPointVlan30, "ipinipd30", 30),
+
+		"ipinipd11": getIPinIPFlow(args, dstEndPointVlan10, "ipinipd11", 11),
+		"ipinipd12": getIPinIPFlow(args, dstEndPointVlan10, "ipinipd12", 12),
+
+		"ipinipd21": getIPinIPFlow(args, dstEndPointVlan20, "ipinipd21", 21),
+		"ipinipd22": getIPinIPFlow(args, dstEndPointVlan20, "ipinipd22", 22),
+
+		"ipinipd31": getIPinIPFlow(args, dstEndPointVlan30, "ipinipd31", 31),
+		"ipinipd32": getIPinIPFlow(args, dstEndPointVlan30, "ipinipd32", 32),
+
+		"ipinipd10v20": getIPinIPFlow(args, dstEndPointVlan20, "ipinipd10v20", 10),
+		"ipinipd11v20": getIPinIPFlow(args, dstEndPointVlan20, "ipinipd11v20", 11),
+		"ipinipd12v20": getIPinIPFlow(args, dstEndPointVlan20, "ipinipd12v20", 12),
+	}
+
+	return f
+}
+
+// getFlows returns a slice of flows based on an input slice of flownames
+func getFlows(f map[string]*ondatra.Flow, keys []string) []*ondatra.Flow {
+	var flows []*ondatra.Flow
+	for _, key := range keys {
+		flows = append(flows, f[key])
+	}
+	return flows
+}
+
+func TestPBR(t *testing.T) {
+	t.Logf("Description: Test RT3.2 with multiple DSCP, IPinIP protocol rule based VRF selection")
+	dut := ondatra.DUT(t, "dut")
+
+	// configure DUT
+	configureDUT(t, dut)
+
+	// Configure ATE
+	ate := ondatra.ATE(t, "ate")
+	top := configureATE(t, ate)
+	top.Push(t).StartProtocols(t)
+
+	args := &testArgs{
+		dut:         dut,
+		ate:         ate,
+		top:         top,
+		srcEndPoint: top.Interfaces()["atePort1"],
+		policyName:  "L3",
+		iptype:      "ipv4",
+		protocol:    oc.PacketMatchTypes_IP_PROTOCOL_IP_IN_IP,
+	}
+
+	allFlows := getAllFlows(args)
+	// dut ingress interface
+	port1 := dut.Port(t, "port1")
+
+	cases := []struct {
+		name         string
+		desc         string
+		policy       *oc.NetworkInstance_PolicyForwarding
+		passingFlows []*ondatra.Flow
+		failingFlows []*ondatra.Flow
+	}{
+		{
+			name: "RT3.2 Case1",
+			desc: "Ensure matching IPinIP with DSCP (10 - VRF10, 20- VRF20, 30-VRF30) traffic reaches appropriate VLAN.",
+			policy: getPBRPolicyForwarding(args,
+				getL3PBRRule(args, "VRF10", 1, []uint8{10}),
+				getL3PBRRule(args, "VRF20", 2, []uint8{20}),
+				getL3PBRRule(args, "VRF30", 3, []uint8{30})),
+			// use IPinIP DSCP10, DSCP20, DSCP30 flows for VLAN10, VLAN20 and VLAN30 respectively.
+			passingFlows: getFlows(allFlows, []string{"ipinipd10", "ipinipd20", "ipinipd30"}),
```

Comment:
Can we write out the individual getIPinIPFlow() calls here? It makes it easier to see how each flow corresponds to the getL3PBRRule in the policy section.

Reply:
thanks @liulk , done